src/replication.c

```diff
@@ -2308,6 +2474,441 @@ char *sendCommandArgv(connection *conn, int argc, char **argv, size_t *argv_lens
+        server.repl_transfer_read = 0;
+        server.repl_transfer_last_fsync_off = 0;
+        server.repl_transfer_lastio = server.unixtime;
+
+        server.repl_rdb_conn_state = REPL_RDB_CONN_RDB_LOAD;
+        return;
+    }
+
+no_response_error:
+    /* Handle receiveSynchronousResponse() error when primary has no reply */
+    serverLog(LL_WARNING, "Primary did not respond to command during SYNC handshake");
+    /* Fall through to regular error handling */
+
+error:
+    connClose(conn);
+    server.repl_transfer_s = NULL;
+    if (server.repl_rdb_transfer_s) {
+        connClose(server.repl_rdb_transfer_s);
+        server.repl_rdb_transfer_s = NULL;
+    }
+    if (server.repl_transfer_fd != -1) close(server.repl_transfer_fd);
+    server.repl_transfer_fd = -1;
+    server.repl_state = REPL_STATE_CONNECT;
+    abortRdbConnectionSync();
+    return;
+
+write_error: /* Handle sendCommand() errors. */
+    serverLog(LL_WARNING, "Sending command to primary in dual connection replication handshake: %s", err);
+    sdsfree(err);
+    goto error;
+}
+
+/* Replication: Replica side.
+ * Initialize server.pending_repl_data infrastructure, we will allocate the buffer
+ * itself once we need it */
+void replDataBufInit(void) {
+    serverAssert(server.pending_repl_data.blocks == NULL);
+    server.pending_repl_data.len = 0;
+    server.pending_repl_data.peak = 0;
+    server.pending_repl_data.blocks = listCreate();
+    server.pending_repl_data.blocks->free = zfree;
+}
+
+/* Replication: Replica side.
+ * Track the local repl-data buffer streaming progress and serve clients from time to time */
+void replStreamProgressCallback(size_t offset, int readlen, time_t *last_progress_callback) {
+    time_t now = mstime();
+    if (server.loading_process_events_interval_bytes &&
+        (offset + readlen) / server.loading_process_events_interval_bytes >
+            offset / server.loading_process_events_interval_bytes &&
+        now - *last_progress_callback > server.loading_process_events_interval_milliseconds) {
+        replicationSendNewlineToPrimary();
+        processEventsWhileBlocked();
+        *last_progress_callback = now;
+    }
+}
+
+/* Replication: Replica side.
+ * Reads replication data from primary into specified repl buffer block */
+int readIntoReplDataBlock(connection *conn, replDataBufBlock *o, size_t read) {
+    int nread = connRead(conn, o->buf + o->used, read);
+    if (nread == -1) {
+        if (connGetState(conn) != CONN_STATE_CONNECTED) {
+            serverLog(LL_VERBOSE, "Error reading from primary: %s", connGetLastError(conn));
+            cancelReplicationHandshake(1);
+        }
+        return C_ERR;
+    }
+    if (nread == 0) {
+        if (server.verbosity <= LL_VERBOSE) {
+            serverLog(LL_VERBOSE, "Provisional primary closed connection");
+        }
+        cancelReplicationHandshake(1);
+        return C_ERR;
+    }
+    o->used += nread;
+    server.stat_total_reads_processed += nread;
```

Comment:
Seems like `stat_total_reads_processed ` is not atomic anymore. Since https://github.com/valkey-io/valkey/commit/bbfd041895ee7d7d8dd59b1c75aa44f5b473e483 we only access it using main thread.

Reply:
Nice!